### PR TITLE
libgit2: fix blatent gc bugs

### PR DESCRIPTION
**Pull request by vtjnash**

### CODE REVIEW

File: base/libgit2/types.jl [hunk 1]
```diff
@@ -210,12 +204,24 @@ Matches the [`git_remote_callbacks`](https://libgit2.github.com/libgit2/#HEAD/ty
     payload::Ptr{Void}
 end
 
-function RemoteCallbacks(credentials_cb::Ptr{Void}, payload::Ref{<:Payload})
```
Comment (ararslan) [inline base/libgit2/types.jl]: I don't think this can be backported unless this non-kwarg method is preserved.

Comment (vtjnash) [reply base/libgit2/types.jl]: I doubt it matters much, but we just leave this part out of the backport

Comment (ararslan) [reply base/libgit2/types.jl]: Then I guess the backport would have to be a separate PR against release-0.6 rather than a cherry pick from this commit?

**Comment by Keno** [comment]
I think this PR will need to be re-done against 0.6. There was different code with different memory safety issues on 0.6. Also, we didn't have gc_preserve.

File: base/libgit2/types.jl (hunk 1 above)
Comment (vtjnash) [reply base/libgit2/types.jl]: Also, fixing this bug requires changing the struct layout (and thus access), which is even more significantly breaking than this method. But like Keno pointed out, the GC was also completely different, so I can probably make a much more conservative fix there that assumes specific missing optimizations.

**Comment by JeffBezanson** [comment]
Maybe there are more? : https://circleci.com/gh/JuliaLang/julia/5681?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Pull request merged**